### PR TITLE
Make adjust dates to context time zone configurable

### PR DESCRIPTION
**Pull request by bauna**
I squashed the changes from [PR-56](https://github.com/nmorel/gwt-jackson/pull/56).
Also addressed your suggestions and added a test for it.

### CODE REVIEW

File: gwt-jackson/src/main/java/com/github/nmorel/gwtjackson/client/utils/DateFormat.java [hunk 1]
```diff
@@ -173,24 +173,30 @@ public static Date parse( String date ) {
...
      *
      * @return the parsed date
      */
-    public static Date parse( String pattern, String date ) {
+    public static Date parse( boolean adjustDatesToContextTimeZone, String pattern, Boolean patternHasTz, String date ) {
         if ( null == pattern ) {
             return parse( DateFormat.DATE_FORMAT_STR_ISO8601, date );
         } else {
-            DateParser parser = CACHE_PARSERS.get( pattern );
+            String patternCacheKey = pattern + adjustDatesToContextTimeZone;
```
Comment (bauna) [inline gwt-jackson/src/main/java/com/github/nmorel/gwtjackson/client/utils/DateFormat.java]: It's necessary to add the adjustDatesToContextTimeZone to the cache key for the case when you use once adjustDatesToContextTimeZone enabled and the in the same app flag disabled.

**Comment by bauna** [comment]
**Don't merge this PR because test DateOptionsJacksonTest is failing.**
After a LOT of search I couldn't found a single use of DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE that seems to be ignored by jackson. Moreover, there are not test using it...
The only config that seems to affects dates deserialization is setting the timezone in the ObjectMapper by the method setTimeZone, and this property by default is set to UTC.
Are you open to make this feature GWT-Jackson feature, maybe change the config feature to something else?

**Comment by nmorel** [comment]
From what I understand from https://github.com/FasterXML/jackson-databind/issues/203 and https://github.com/FasterXML/jackson-databind/issues/204, the feature is only available for joda and is different from what you want I think.

If set to false, date without a timezone will still use the context timezone, the one in `ObjectMapper`.
It only affects date *with* a timezone. Set to true or false, the milliseconds to epoch will still be the same. Only the timezone is affected. Instead of having UTC timezone (the one in `ObjectMapper`), you have a date with the timezone defined in the string.

I think what you want is more an option to change the context timezone, a `useBrowserTimezone` configuration. If set to true, you don't use the ` +0000` trick to parse to UTC when there is no timezone.
And to test that, in jackson, you set the timezone in the ObjectMapper to the current one.

Maybe it could also be used in `JsonFormat#timezone`. Currently, I don't support it because GWT only have timezone offset unless you load all the locales. But since it takes a string, we could have a constant like `browser` that tells us to use the browser timezone instead of the default one which is UTC.

**Comment by bauna** [comment]
Thanks for the feedback. I'll implement your first suggestion because it plays better for my case where I'm using gwt-jackson from gwtp-rest-dispatch.
And the second idea of setting an special string un JsonFormat#timezone has the problem that won't work when you want to use the same classes in client and server.

**Comment by bauna** [comment]
closed see [PR-61](https://github.com/nmorel/gwt-jackson/pull/61)

**Pull request closed without merging**